### PR TITLE
List any kubernetes resource based on available api-resources

### PR DESCRIPTION
**Pull request by ifitzpat**
Hi and thank you for the wonderful work on kubel (and very informative talk)!

I made some additions to allow the user to list any kubernetes resource based on resources that are available in whatever context they're working in. The code is very much a work-in-progress (hacky calls to "awk", terse documentation, etc.), but I wanted to create the pull-request at this stage to see if the functionality might be of wider use. If you're interested in this kind of feature, I can spend some time refactoring before you merge.

Thanks again!

### CODE REVIEW

**Comment by abrochard** [comment]
Thank you very much! I am glad it is interesting to anyone.

I did a quick skim of your code and I am a bit confused. My understanding is that you expand from the basic tabulated list of pods into a more generic "resource" view where one can select the particular resource with the `R` key. Is that correct?

I need to play around a bit with it but it is pretty cool. Here are my areas of concerns:
- `kubectl api-resources` seems to be a 1.13.3 feature, unfortunately that version isn't supported for me yet
- is this an smart UX? In my mind, the default pod view is really more a placeholder/dashboard. It shows you k8s stuff, but the real action happens with deployments and services which use the `completing-read` options for more interactive usage
- is this going to break other features? if I select `configmaps` as a resource and press enter on one of them, what is going to happen?

In any case, I like it. I will test it out.

**Comment by ifitzpat** [comment]
Thanks for your feedback!

> I did a quick skim of your code and I am a bit confused. My understanding is that you expand from  the basic tabulated list of pods into a more generic "resource" view where one can select the particular resource with the R key. Is that correct?

Yes that's exactly right

> kubectl api-resources seems to be a 1.13.3 feature, unfortunately that version isn't supported for me yet

Right! I'll see if I can add some exception handling to only go with the "full list" of resources when version is >=1.13.3 and default to a more limited list otherwise

> is this an smart UX? In my mind, the default pod view is really more a placeholder/dashboard. It shows you k8s stuff, but the real action happens with deployments and services which use the completing-read options for more interactive usage.

To me it's valuable to be able to list different resource types and drill down from there. It's fair to say I'm no UX specialist, though ;-)

> I added a commit to remedy this behaviour. (RET) calls "describe" on whichever resource type is current. None of the original functionality should be compromised.

Thanks again and I look forward to any further feedback you might have!

**Comment by abrochard** [comment]
Sounds great! Let's get this in shape.

**Comment by abrochard** [comment]
Just played around with this. I really like it. Would love to merge it. I think the main outstanding points are probably:

- cleaning up the code (removing awk stuff)
- implementing `kubel-get-resource-details`
- supporting a downgraded experience when `kubectl api-resources` isn't available. Maybe doing a `completing-read` with a list of hard coded common resources
- fixing the color display for the pod view, and maybe even providing a meaningful colorization for some of the other view

**Comment by ifitzpat** [comment]
Agreed on all fronts. I wrote kubel-get-resource-details but it seems I hadn't pushed it yet.

**Comment by ifitzpat** [comment]
**Still TODO:**

- [ ] cleaning up the code (removing awk stuff)
- [ ] fixing the color display for the pod view, and maybe even providing a meaningful colorization for some of the other view

File: kubel.el [hunk 1]
```diff
@@ -372,6 +372,15 @@ TYPENAME is the resource type/name."
     (kubel--exec buffer-name nil (list "describe" "pod" (kubel--get-pod-under-cursor)))
     (beginning-of-buffer)))
 
+(defun kubel-get-resource-details ()
+  "Get the details of the resource under the cursor."
+  (interactive)
+  (let* ((pod (kubel--get-pod-under-cursor))
```
Comment (abrochard) [inline kubel.el]: would it be worth renaming `kubel--get-pod-under-cursor` to something like `kubel--get-resource-under-cursor` ?

Comment (ifitzpat) [reply kubel.el]: Done! Will test for regressions.

**Comment by ifitzpat** [comment]
Not sure what you will make of the last few commits. It seemed to me that the various kubel-describe functions are redundant with the introduction of the kubel-get-resource-details function. To preserve the option of yielding yaml I bound <RET> tot kubel-describe-popup. The downside is you now need to do <RET> <RET> to view details...

**Comment by abrochard** [comment]
I've been thinking about this a little and I think that although it makes sense from an visual representation perspective, I would like to spend a bit more time considering the UX that this PR will introduce.

Maybe it would make sense to tag the current version of Kubel as 1.0, and create a new `develop` branch to merge your changes in. From there I can contribute and we can work on polishing. It would even be ideal to find other Kubel users to ask for their opinion, but I have no idea how to do that. Would that work for you?

Here are my thoughts currently on this. Right now the "default use case" I built Kubel for is:
1. summon kubel
2. see pods
3. tail logs / perform pod relevant actions with a meaningful transient
4. use a clunky system to edit a deployment
5. mash the `g` key until I see the new pods come up

If I were to decompose the clunky system of (4) I would get:
1. press `d`
2. press `-y` to switch to yaml
3. press `d` again
4. enter with completing-read the deployment name
5. press enter
6. make changes
7. `C-c C-c` to apply

With your changes, this step would be:
1. press `R` 
2. select `deployment` as resource with a completing-read
3. `C-s` and enter the deployment name to jump to the line
4. press enter
5. press `-y` to switch to yaml
6. press enter again
7. make changes
8. `C-c C-c` to apply

I think the first evident thing is that we should get rid of that `-y` option and make yaml default. I always use yaml and I think that most people would. We can always leave a flag/option to make the default non-yaml. Ideally, `C-c C-c` would should you the diff and ask you to confirm before applying too, but that's another story.

The second issue, is that after all these steps I have to switch back to the pod views to see when my changes are up. I think that's something I need to test.

Finally, what I really need to dig into is making sure that the transient/keybindings are relevant to the current view. If I am looking at services, I should not have the option to see the log transient or exec into a container. I wonder how magit does it, how they have relevant options dependent on where you click and which view you are in. 

Sorry for the novel, in short, I propose that PR be made against a separate branch, merged, and polished.

**Comment by ifitzpat** [comment]
No worries! I welcome the discussion and I think a user consultation makes sense! I don't think you need a different PR/develop branch to contribute though. You should be able to commit to my master branch already (it's an option on the PR). Keeping it in this PR also retains all the comments/history.

> I think the first evident thing is that we should get rid of that -y option and make yaml default. I always use yaml and I think that most people would. We can always leave a flag/option to make the default non-yaml. 

Agreed! We could make this a defcustom?

> Ideally, C-c C-c would should you the diff and ask you to confirm before applying too, but that's another story.

Yup this should probably be a different feature branch.

> The second issue, is that after all these steps I have to switch back to the pod views to see when my changes are up. I think that's something I need to test.

Ok, maybe bury the buffer or automate switching back? Also feels like it should be another branch to me.

> Finally, what I really need to dig into is making sure that the transient/keybindings are relevant to the current view. If I am looking at services, I should not have the option to see the log transient or exec into a container. I wonder how magit does it, how they have relevant options dependent on where you click and which view you are in.

Yes absolutely! I think we should tackle this in this PR. Possibly populate the popup after checking kubel--resource?

My kubel usage is slightly different from yours. I have an orgmode file with a literate config of my k8s resources. I use [ob-kubectl](https://github.com/ifitzpat/ob-kubectl) to apply changes and check the status/config of different resource types using kubel. I used to use kubernetes.el to do all this, but I found it difficult to extend to non-"standard" resource types (e.g., KNative CRDs).

**Comment by ifitzpat** [comment]
Should we invite user comments to this PR (maybe advertise on Reddit/Telegram/IRC)?

**Comment by ifitzpat** [comment]
Come to think of it: we could support both workflows?

**Comment by abrochard** [comment]
> Should we invite user comments to this PR (maybe advertise on Reddit/Telegram/IRC)?

I guess we could? I always feel a bit obnoxious.
Maybe I could start by asking a past contributor.

@snowiow @maplemuse: do you guys want to chime in? what is your kubel workflow. would you be negatively affected?

**Comment by abrochard** [comment]
> No worries! I welcome the discussion and I think a user consultation makes sense! I don't think you need a different PR/develop branch to contribute though. You should be able to commit to my master branch already (it's an option on the PR). Keeping it in this PR also retains all the comments/history.
So fancy, I'll try that.


> Agreed! We could make this a defcustom?
 👍 

> > The second issue, is that after all these steps I have to switch back to the pod views to see when my changes are up. I think that's something I need to test.
> 
> Ok, maybe bury the buffer or automate switching back? Also feels like it should be another branch to me.
👍 


> My kubel usage is slightly different from yours. I have an orgmode file with a literate config of my k8s resources. I use [ob-kubectl](https://github.com/ifitzpat/ob-kubectl) to apply changes and check the status/config of different resource types using kubel. I used to use kubernetes.el to do all this, but I found it difficult to extend to non-"standard" resource types (e.g., KNative CRDs).
Ah that makes total sense.

>>Finally, what I really need to dig into is making sure that the transient/keybindings are relevant to the current view. If I am looking at services, I should not have the option to see the log transient or exec into a container. I wonder how magit does it, how they have relevant options dependent on where you click and which view you are in.
>Yes absolutely! I think we should tackle this in this PR. Possibly populate the popup after checking kubel--resource?
Ok I will try to play with the transient a bit

**Comment by ifitzpat** [comment]
TODO: Deal with missing fields e.g.,
```
NAME            URL                                          LATESTCREATED         LATESTREADY           READY     REASON
busybox-test    http://busybox-test.default.cluster.local    busybox-test-xlnj2                          Unknown   RevisionMissing
helloworld-go   http://helloworld-go.default.cluster.local   helloworld-go-rkqjv   helloworld-go-rkqjv   True
```

**Comment by ifitzpat** [comment]
>>> Finally, what I really need to dig into is making sure that the transient/keybindings are relevant to the current view. If I am looking at services, I should not have the option to see the log transient or exec into a container. I wonder how magit does it, how they have relevant options dependent on where you click and which view you are in.

>> Yes absolutely! I think we should tackle this in this PR. Possibly populate the popup after checking kubel--resource?

> Ok I will try to play with the transient a bit

it looks like there are options of including control-logic in the transiets. This is something I found in magit-log.el, for example:

```
 [:if-non-nil magit-wip-mode
    :description "Wiplog"
    ("i" "index"          magit-wip-log-index)
    ("w" "worktree"       magit-wip-log-worktree)]])

```

**Comment by abrochard** [comment]
> it looks like there are options of including control-logic in the transiets. This is something I found in magit-log.el, for example:
> 
> ```
>  [:if-non-nil magit-wip-mode
>     :description "Wiplog"
>     ("i" "index"          magit-wip-log-index)
>     ("w" "worktree"       magit-wip-log-worktree)]])
> ```

Amazing, just pushed something making use of that.

**Comment by abrochard** [comment]
Actually I think I'm debating this conditional transient thing. Rather than have keybindings specific to a particular view (like `p` to port-forward a pod), maybe we could do a keybinding with slightly different behavior based on your view:
- if you are in the pod view, `p` will pod-forward what's under the cursor
- if you are not in the pod view, `p` will popup a completing-read prompt with a pod list

That way we actually don't break existing workflows and nicely integrate with more views.
I'm also thinking about adding a "quick edit" type command which would work similarly.

**Comment by ifitzpat** [comment]
> 
    if you are in the pod view, p will pod-forward what's under the cursor
    if you are not in the pod view, p will popup a completing-read prompt with a pod list

That makes a lot of sense and might actually be easier to implement!

**Comment by abrochard** [comment]
Great, I just pushed that

**Comment by abrochard** [comment]
Ok I think I made the last functional change. 
What's left before we merge:
- review the whole code to make sure naming/formatting makes sense and track unused functions

What's left after merge:
- update changelog
- update doc
- tag version 2.0

Anything left for you?

**Comment by ifitzpat** [comment]
> Anything left for you?
This missing fields bug (above) is the most important I think. I'll have a think about how kubel--parse-body can deal with that. What I can think about at the moment is a check that each list contains an equal number of members... not sure...

**Comment by abrochard** [comment]
> > Anything left for you?
> 
> This missing fields bug (above) is the most important I think. I'll have a think about how kubel--parse-body can deal with that. What I can think about at the moment is a check that each list contains an equal number of members... not sure...
> 
> EDIT: formatting

What's the root cause? How do you reproduce?

**Comment by ifitzpat** [comment]
Ok, the missing field bug seems to be solved by making the very hacky and ugly assumption that missing values will be represented by a sequence of at leat 25 spaces. On my cluster this works for all resource types, but the assumption is easily violated...

**Comment by ifitzpat** [comment]
> What's the root cause? How do you reproduce?
Take the following example:

```
NAME            URL                                          LATESTCREATED         LATESTREADY           READY     REASON
busybox-test    http://busybox-test.default.cluster.local    busybox-test-xlnj2                          Unknown   RevisionMissing
helloworld-go   http://helloworld-go.default.cluster.local   helloworld-go-rkqjv   helloworld-go-rkqjv   True
```
in kubel--parse-body we first replaces every sequence one space with a single space and then called split-string. In the above example the missing value in the LATEST READY column gets dropped and the line is output with a missing column.

**Comment by abrochard** [comment]
Ah I see. Unfortunately that hack is completely messing up on my cluster. In my pod view, the status ends up in the restart column. There must be a better way.
The long term fix is to stop using the vanilla `get [resource]` and instead parse the JSON but it seems hard now...

**Comment by ifitzpat** [comment]
> Ah I see. Unfortunately that hack is completely messing up on my cluster. In my pod view, the status ends up in the restart column. There must be a better way.

 :( ok, there's a more verbose way I suppose. Parse the header row separately. Save the number of columns and provide that to kubel--parse line. If the number of extracted columns is lower than the number of columns in the header then substitute the longest sequence of spaces with " - "... 

> The long term fix is to stop using the vanilla get [resource] and instead parse the JSON but it seems hard now...

Yeah that would solve the missing value issue, but it might be difficult to define which fields to return from the json without having to hard-code them for each resource type.

**Comment by abrochard** [comment]
I just pushed a complicated fix for it. It uses the header line to determine the bounds of each column and then parses each line using those bounds. Let me know if it makes sense and if there's a better way to write it.

**Comment by abrochard** [comment]
@ifitzpat are you ok with me merging this now and finishing up the last doc changes on my side?

**Comment by ifitzpat** [comment]
Hi @abrochard, sorry for the radio silence! I checked out the new code today and the only issue remaining is that the header row parsing doesn't seem to deal with spaces in headers. (eg., "LATEST READY       READY" becomes "LATEST    READYREADY"). It looks like the first opportunity I'll have to do any work on this issue will be this weekend. In the meantime I'm fine with merging and maybe flagging this as a known issue.

**Comment by abrochard** [comment]
No worries at all @ifitzpat . I just pushed what I think is a fix? Somebody is going to have a blast making unit test for all of these...
I will merge today and please let me know if you see anything else.
Thank you so much for your help!

**Pull request merged**